src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+#else
+
+  std::vector<struct pollfd> pollfds;
+  pollfds.reserve(args->num_fds);
+  struct pollfd pfd;
+  for (int i = 0; i < args->rfds; i++) {
+    pfd.fd = fds[i];
+    pfd.events = POLLIN;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+    pfd.fd = fds[i];
+    pfd.events = POLLOUT;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+    pfd.fd = fds[i];
+    pfd.events = 0;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+
+  do {
+    result = POLL_FUNC(pollfds.data(), args->num_fds, repeats ? LATER_INTERVAL : timeoutms);
+    if (args->flag->load()) goto callback;
+    if (result) break;
+  } while (infinite || (repeats-- && (timeoutms -= LATER_INTERVAL)));
+
+  if (result > 0) {
+    for (int i = 0; i < args->rfds; i++) {
+      fds[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLIN ? 1: R_NaInt;
+    }
+    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+      fds[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLOUT ? 1 : R_NaInt;
+    }
+    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+      fds[i] = pollfds[i].revents != 0;
+    }
+  } else if (result == 0) {
+    std::memset(fds, 0, args->num_fds * sizeof(int));
+  } else {
+    for (int i = 0; i < args->num_fds; i++) {
+      fds[i] = R_NaInt;
+    }
+  }
+
+#endif // POLLIN
+
+  callback:
+  callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
+
+  return 0;
+
+}
+
+// [[Rcpp::export]]
+Rcpp::RObject execLater_fd(Rcpp::Function callback, Rcpp::IntegerVector readfds, Rcpp::IntegerVector writefds,
+                                 Rcpp::IntegerVector exceptfds, Rcpp::NumericVector timeoutSecs, Rcpp::IntegerVector loop_id) {
+
+  const int rfds = static_cast<int>(readfds.size());
+  const int wfds = static_cast<int>(writefds.size());
+  const int efds = static_cast<int>(exceptfds.size());
+  const int num_fds = rfds + wfds + efds;
+  if (num_fds == 0)
+    Rcpp::stop("No file descriptors supplied");
+
+  std::shared_ptr<ThreadArgs> args = std::make_shared<ThreadArgs>();
+  std::unique_ptr<std::vector<int>> fds(new std::vector<int>());
+  fds->reserve(num_fds);
+  std::shared_ptr<std::atomic<bool>> flag = std::make_shared<std::atomic<bool>>();
+  args->flag = flag;
+  args->timeout = timeoutSecs[0];
+  args->loop = loop_id[0];
+  args->num_fds = num_fds;
+  args->rfds = rfds;
+  args->wfds = wfds;
+  args->efds = efds;
+  for (int i = 0; i < rfds; i++) {
+    fds->push_back(readfds[i]);
+  }
+  for (int i = 0; i < wfds; i++) {
+    fds->push_back(writefds[i]);
+  }
+  for (int i = 0; i < efds; i++) {
+    fds->push_back(exceptfds[i]);
+  }
+  args->fds = std::move(fds);
+
+  SEXP call = Rf_lcons(callback, Rf_cons(Rf_allocVector(LGLSXP, num_fds), R_NilValue));
+  R_PreserveObject(call);
```

Comment:
In section 5.9.1 of https://cran.r-project.org/doc/manuals/R-exts.html it says to only use this "sparingly" and in the past I've taken that pretty seriously. For example calling `R_PreserveObject` on just one environment for the whole package and keeping our precious objects in that.

But also, is there a reason you're creating this `call` object here, instead of storing the Rcpp::Function object directly on the `ThreadArgs` object, and then calling it with regular Rcpp/C++ function calling instead of `Rf_Eval`?

Reply:
Agree with sparing use of `R_PreserveObject()`. I maintain my own precious list in nanonext (much like Rcpp does) where there can be thousands/millions of concurrent messages, but here I didn't expect there to be more than a handful of waits at the same time.

In terms of creating the call, it's just what I'm used to in terms of getting maximum efficiency, and I suspect what Rcpp ultimately does under layers of sugar (as this is an arbitrary R call, not an entry point into compiled code) - I could be wrong.

However your idea for using Rcpp function calling is a good one and I've made the switch in [852e1d0](https://github.com/r-lib/later/pull/190/commits/852e1d07c14430b1a5d0c139fb54c7024c9f0e2d) as:
1. Allows me to manage the object lifetime with a `std::unique_ptr`.
2. Makes the `R_PreserveObject()` point moot - no need to maintain our own precious list.
3. Cancellation can now return early from the background thread without needing to trigger the later callback at all.